iroh-bytes/src/provider.rs

```diff
@@ -616,8 +734,107 @@ pub async fn send_blob<D: BaoMap, W: AsyncWrite + Unpin + Send + 'static>(
             Ok((SentStatus::Sent, size))
         }
         _ => {
-            debug!("blob not found {}", name);
+            debug!("blob not found {}", hex::encode(name));
             Ok((SentStatus::NotFound, 0))
         }
     }
 }
+
+/// The mutable part of a BaoDb
+pub trait BaoDb: BaoReadonlyDb + BaoPartialMap {
+    /// list partial blobs in the database
+    fn partial_blobs(&self) -> Box<dyn Iterator<Item = Hash> + Send + Sync + 'static>;
+
+    /// extract a file to a local path
+    ///
+    /// `hash` is the hash of the file
+    /// `target` is the path to the target file
+    /// `stable` is true if the file can be assumed to be retained unchanged in the file system
+    /// `progress` is a callback that is called with the total number of bytes that have been written
+    fn export(
+        &self,
+        hash: Hash,
+        target: PathBuf,
+        stable: bool,
+        progress: impl Fn(u64) -> io::Result<()> + Send + Sync + 'static,
+    ) -> BoxFuture<'_, io::Result<()>>;
+
+    /// import a file from a local path
+    ///
+    /// `data` is the path to the file
+    /// `stable` is true if the file can be assumed to be retained unchanged in the file system. If
+    /// `stable` is false, the file will be copied.
```

Comment:
I wonder if `should_copy` would be a better name

Reply:
I think I will just make an enum...